Add a small tutorial on using Resource scripts.

willnationsdev opened the pull request:
I can't vouch (at all) for the C# snippets in here, so someone more familiar with Godot's C# syntax will have to proofread those. I'm also not sure if my links are executed properly and whether I can safely nest code snippets within a note or warning (as I've done here).

This should finally help people get a better idea of how useful Resource scripts can be, how to make them, and how to use them.

mhilbrunner commented:
Thanks for this! I'll proofread - FYI, building the docs yourself locally is a great way to verify how it looks in the output, and isn't too hard: Look at the [Building with Sphinx](https://github.com/godotengine/godot-docs#building-with-sphinx) section of the README in this repository.

willnationsdev commented:
@mhilbrunner I got it rendering locally (yay!), so I've fixed up a bunch of stuff with the syntax, however I can't seem to figure out how I'm supposed to make the "script classes" link work. It's supposed to link to the "Scripting Continued" page at the "Register scripts as classes" header, but all the variations I've tried (like :ref:\`script classes <scripting_continued#register-scripts-as-classes>\`) don't seem to work.

LikeLakers2 commented on getting_started/step_by_step/resources.rst:
```diff
@@ -158,12 +158,173 @@ Freeing resources
+
+- Resources can even serialize sub-Resources recursively, meaning users can design even more sophisticated data structures.
+
+- Users can save Resources as version-control-friendly text files (\*.tres). Upon exporting a game, Godot serializes resource files as binary files (\*.res) for increased compression.
```
Aren't `*.res` files also faster to load?

willnationsdev replied:
Yes, because the same amount of data is packed into a smaller space, so it doesn't take as long to parse the data as input.

willnationsdev replied:
I suppose I could do "for increased speed and compression" instead.

LikeLakers2 commented on getting_started/step_by_step/resources.rst:
```diff
@@ -158,12 +158,173 @@ Freeing resources
+    func _ready():
+        print(stats.health) # prints '10'
+  ..code-tab:: csharp
+        using Godot;
```
I think you forgot to add a C# version of `bot_stats.gd`.

willnationsdev replied:
Oh yeah, guess I could add that. Like I said though, I doubt any of the C# code is really accurate since I haven't used Godot C# at all.

LikeLakers2 replied:
Since you added a C# version, mind resolving this conversation? I can't seem to do so myself.

LikeLakers2 commented on getting_started/step_by_step/resources.rst:
```diff
@@ -158,12 +158,173 @@ Freeing resources
+    With GDScript, it only supports Resource classes at the top level. Defining
+    a Resource subclass, i.e. an inner class, within a GDScript file will have no 
+    effect. If one attempts to save an instance of a Resource subclass, none of 
+    the scripted properties will serialize to the saved file.
```
This sounds like a bug to me -- I've saved `ShaderMaterial`s with `Shader` resources inside them before. I don't see why this should be any different.

willnationsdev replied:
You can have sub resources within scripts, but if you try to define an entirely new type of Resource *as a subclass within another script*, then Godot isn't going to identify the path for the script on that resource as "path to the file that has the Resource subclass AND somehow mark that you need to assign the constant value of the subclass Script resource as the script value for the created \*.tres. You know, something like "res://file.gd#MyResource" or what-have-you.

willnationsdev replied:
Like, if I make a Resource object and set it's script to be not another script file, but a Resource class that's a subclass of a script file (where the script file could derive ANYTHING), it won't start serializing those script properties properly or anything.

LikeLakers2 replied:
Thing is, the way this issue is being described in this warning makes it come off as a bug. It sounds like something Godot *could* do, and that it probably should do, but for one reason or another just... doesn't do.

Maybe try to explain why this is the case, at least?

willnationsdev replied:
Sure, sounds good.

williamd1k0 commented on getting_started/step_by_step/resources.rst:
```diff
@@ -158,12 +158,173 @@ Freeing resources
+.. note::
+
+   Resource scripts are similar to Unity's ScriptableObjects. However, unlike in
+   Unity, there is no need to implement IMGUI logic to enable users to create
```
In fact, there is no need to implement immediate gui or custom editor to use a "common" ScriptableObject.

The current workflow (I tested it in Unity 2018.3):

1. Create your `ScriptableObject` class:
```c#
public class CustomData : ScriptableObject {
    public string data;
}
```
2. Add it to `Assets->Create` menu:
```c#
// This is equivalent to `class_name` in Godot 3.1+
[CreateAssetMenu(menuName="MyCustomStuff/CustomData")]
public class CustomData : ScriptableObject {
    public string data;
}
```
3. Create a component that will use your `ScriptableObject`:
```c#
public class MyComponent : MonoBehaviour {
    public CustomData data;
}
```
4. Create your asset using the `Assets->Create` menu;
5. Drag and drop it to the `data` field in the inspector.

willnationsdev replied:
Oh, interesting. That's news to me. Great that they added that feature! Appreciate the heads-up. I'll go ahead and take out that tidbit then.

williamd1k0 replied:
Unity docs and changelogs are pretty confusing so I don't know in which version this started to work this way :sweat_smile:

willnationsdev commented:
@williamd1k0 @LikeLakers2 Updated the document to reflect all of the changes.

willnationsdev commented:
Okay, finally figured out how to make the script classes link work (just doing an absolute link via URL rather than relative one with :ref:).

mhilbrunner commented on getting_started/step_by_step/resources.rst:
```diff
@@ -158,12 +158,201 @@ Freeing resources
+scripts grant all the same benefits as the base Resource object. They
+inherit both reference-counting memory management and automatic serialization
+between binary or text data (\*.res, \*.tres) and in-memory object
+properties.
```
*Resource scripts grant all the same benefits as the base Resource object.
They inherit both reference-counting memory management and automatic serialization
between binary or text data (\*.res, \*.tres) and in-memory object properties.*

This is IMO very complex. We need to avoid overwhelming users with cognitive overload :)

How about:

*Resource scripts inherit all the same benefits as the base Resource object:
Automatic memory management using reference counting, automatic serialization
between binary or text data (\*.res, \*.tres) and in-memory object properties.*

exts commented on getting_started/step_by_step/resources.rst:
```diff
@@ -158,12 +158,201 @@ Freeing resources
+            public int health;
+
+            [Export]
+            public Resource subResource;
```
http://docs.godotengine.org/en/latest/getting_started/scripting/c_sharp/c_sharp_style_guide.html#naming-conventions

public methods/properties need to be PascalCase. Update this for every property.

exts commented on getting_started/step_by_step/resources.rst:
```diff
@@ -158,12 +158,201 @@ Freeing resources
+
+        public class BotStatsTable : Resource
+        {
+            private Godot.Dictionary<String, BotStats> stats = new Godot.Dictionary<String, BotStats>();
```
Use underscore prefix for private members http://docs.godotengine.org/en/latest/getting_started/scripting/c_sharp/c_sharp_style_guide.html#naming-conventions

exts commented on getting_started/step_by_step/resources.rst:
```diff
@@ -158,12 +158,201 @@ Freeing resources
+        public class Bot : KinematicBody
+        {
+            [Export]
+            public BotStats stats;
```
change to PascalCase http://docs.godotengine.org/en/latest/getting_started/scripting/c_sharp/c_sharp_style_guide.html#naming-conventions

exts commented on getting_started/step_by_step/resources.rst:
```diff
@@ -158,12 +158,201 @@ Freeing resources
+            public Resource subResource;
+
+            [Export]
+            public String[] strArray;
```
Change to PascalCase http://docs.godotengine.org/en/latest/getting_started/scripting/c_sharp/c_sharp_style_guide.html#naming-conventions

exts commented on getting_started/step_by_step/resources.rst:
```diff
@@ -158,12 +158,201 @@ Freeing resources
+        public class BotStats : Resource
+        {
+            [Export]
+            public int health;
```
Change to PascalCase http://docs.godotengine.org/en/latest/getting_started/scripting/c_sharp/c_sharp_style_guide.html#naming-conventions

willnationsdev commented:
@mhilbrunner @exts updated to reflect the syntax changes and the rewording of the introductory paragraph. I have also added an additional warning and code segment for GDScript (not even sure if a C# example would apply for the warning since it has to do with Dictionary vs. Resource memory management and I don't know if C# GC's Dictionaries). I also fixed a few rendering errors that were preventing the examples from showing up.

willnationsdev commented:
Oh, also, I realized that I've been mixing and matching my C# code examples' syntax for allocating objects, but I don't see anything in the docs to indicate what it looks like in the style guide or API differences. Which of these would be used?

    MyResource res = new MyResource(10);
    MyResource res = MyResource.new(10);

KellyThomas commented on getting_started/step_by_step/resources.rst:
```diff
@@ -158,12 +158,259 @@ Freeing resources
+
+                public override void _Init()
+                {
+                    if (!Stats)
```
To perform a null check we should use:

    if (stats == null)

KellyThomas commented on getting_started/step_by_step/resources.rst:
```diff
@@ -158,12 +158,259 @@ Freeing resources
+                {
+                    if (!Stats)
+                    {
+                        Stats = (ResourceLoader::load("BotStats.cs") as BotStats).new(10);
```
If we are using the constructor as suggested above we can use:

    stats = new BotStats(10);

KellyThomas commented on getting_started/step_by_step/resources.rst:
```diff
@@ -158,12 +158,259 @@ Freeing resources
+                [Export]
+                public BotStats Stats;
+
+                public override void _Init()
```
While testing this code I found that `_Init()` isn't being called (https://github.com/godotengine/godot/issues/22633).

KellyThomas commented on getting_started/step_by_step/resources.rst:
```diff
@@ -158,12 +158,259 @@ Freeing resources
+    
+    func _ready():
+        print(stats.health) # prints '10'
+  .. code-tab:: csharp
```
Is this code-tab's code indented more than the others?

willnationsdev replied:
Should be fixed now.

willnationsdev commented on getting_started/step_by_step/resources.rst:
```diff
@@ -158,12 +158,259 @@ Freeing resources
+            public class Bot : KinematicBody
+            {
+                [Export]
+                public BotStats Stats;
```
If there isn't a way to export the type (scripted types may not be supported this way?), perhaps something where we export `Resource`, but then do a runtime check that the type extends `BotStats`.

KellyThomas replied:
BotStats is not displayed in the Create New Resource dialog either.
Its unclear to me how to create one using the editor GUI.

willnationsdev replied:
Yeah, this seems like a deeper issue than I realized. https://github.com/godotengine/godot/issues/22641

willnationsdev replied:
As for right now, you basically HAVE to export it as an engine type (Resource or some derivative that is still in the engine) and then check if the runtime instance is a script at runtime.

willnationsdev commented:
Updated to reflect all changes and relevant Issues.

KellyThomas commented on getting_started/step_by_step/resources.rst:
```diff
@@ -158,12 +158,260 @@ Freeing resources
+        public class MyNode : Node
+        {
+
+            public class MyResource : Resource {
```
Its all looking good now, except this brace would be best on the next line.

KellyThomas commented on getting_started/step_by_step/resources.rst:
```diff
@@ -158,12 +158,260 @@ Freeing resources
+                public override void _Ready()
+                {
+                    if (Stats != null && Stats is BotStats) {
+                        Godot.print((Stats as BotStats).Health);
```
Also this should be `GD.Print()`.

NathanLovato commented:
Thanks much for your work!

NathanLovato commented:
Sorry, just added the two small changes!
